integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -0,0 +1,633 @@
+      LOG.debug("Failed to get info of {}, path does not exist or is invalid", path);
+      return -ErrorCodes.ENOENT();
+    } catch (Throwable e) {
+      LOG.error("Failed to getattr {}: ", path, e);
+      return -ErrorCodes.EIO();
+    }
+
+    return 0;
+  }
+
+  @Override
+  public int readdir(String path, long buff, FuseFillDir filter, long offset,
+      FuseFileInfo fi) {
+    return AlluxioFuseUtils.call(LOG, () -> readdirInternal(path, buff, filter, offset, fi),
+        "readdir", "path=%s,buf=%s", path, buff);
+  }
+
+  private int readdirInternal(String path, long buff, FuseFillDir filter, long offset,
+      FuseFileInfo fi) {
+    final AlluxioURI uri = mPathResolverCache.getUnchecked(path);
+    try {
+      // standard . and .. entries
+      filter.apply(buff, ".", null, 0);
+      filter.apply(buff, "..", null, 0);
+
+      mFileSystem.iterateStatus(uri, file -> {
+        filter.apply(buff, file.getName(), null, 0);
+      });
+    } catch (Throwable e) {
+      LOG.error("Failed to readdir {}: ", path, e);
+      return -ErrorCodes.EIO();
+    }
+
+    return 0;
+  }
+
+  @Override
+  public int open(String path, FuseFileInfo fi) {
+    return AlluxioFuseUtils.call(LOG, () -> openInternal(path, fi), "open", "path=%s", path);
+  }
+
+  private int openInternal(String path, FuseFileInfo fi) {
+    final AlluxioURI uri = mPathResolverCache.getUnchecked(path);
+    try {
+      long fd = mNextOpenFileId.getAndIncrement();
+      FileInStream is = mFileSystem.openFile(uri);
+      mOpenFileEntries.put(fd, is);
+      fi.fh.set(fd);
+      return 0;
+    } catch (Throwable e) {
+      LOG.error("Failed to open {}: ", path, e);
+      return -ErrorCodes.EIO();
+    }
+  }
+
+  @Override
+  public int read(String path, ByteBuffer buf, long size, long offset, FuseFileInfo fi) {
+    return AlluxioFuseUtils.call(LOG, () -> readInternal(path, buf, size, offset, fi),
+        "read", "path=%s,buf=%s,size=%d,offset=%d", path, buf, size, offset);
+  }
+
+  private int readInternal(String path, ByteBuffer buf, long size, long offset, FuseFileInfo fi) {
+    int nread = 0;
+    int rd = 0;
+    final int sz = (int) size;
+    long fd = fi.fh.get();
+    // FileInStream is not thread safe
+    try (LockResource r1 = new LockResource(getFileLock(fd).writeLock())) {
+      FileInStream is = mOpenFileEntries.get(fd);
+      if (is == null) {
+        LOG.error("Cannot find fd {} for {}", fd, path);
+        return -ErrorCodes.EBADFD();
+      }
+      is.seek(offset);
+      final byte[] dest = new byte[sz];
+      while (rd >= 0 && nread < size) {
+        rd = is.read(dest, nread, sz - nread);
+        if (rd >= 0) {
+          nread += rd;
+        }
+      }
+
+      if (nread == -1) { // EOF
+        nread = 0;
+      } else if (nread > 0) {
+        buf.put(dest, 0, nread);
+      }
+    } catch (Throwable e) {
+      LOG.error("Failed to read {},{},{}: ", path, size, offset, e);
```

Comment:
Can we add labels to the log, like "path: {}, size: {}, offset: {}"?

Reply:
updated